# Assignment 2: Part 1
Hallee Pham

## Setup
* Imports
* Read data

In [20]:
import pandas as pd
import numpy as np

# Read data
cars = pd.read_csv('https://raw.githubusercontent.com/halleepham/DataScience-Assignment2/refs/heads/main/Part1/data/raw/train.csv', index_col=0)
cars.head(15)

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67 kmpl,1582 CC,126.2 bhp,5.0,NaN,12.50
2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,13 km/kg,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77 kmpl,1248 CC,88.76 bhp,7.0,NaN,6.00
4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.2 kmpl,1968 CC,140.8 bhp,5.0,NaN,17.74
6,Nissan Micra Diesel XV,Jaipur,2013,86999,Diesel,Manual,First,23.08 kmpl,1461 CC,63.1 bhp,5.0,NaN,3.50
7,Toyota Innova Crysta 2.8 GX AT 8S,Mumbai,2016,36000,Diesel,Automatic,First,11.36 kmpl,2755 CC,171.5 bhp,8.0,21 Lakh,17.50
8,Volkswagen Vento Diesel Comfortline,Pune,2013,64430,Diesel,Manual,First,20.54 kmpl,1598 CC,103.6 bhp,5.0,NaN,5.20
9,Tata Indica Vista Quadrajet LS,Chennai,2012,65932,Diesel,Manual,Second,22.3 kmpl,1248 CC,74 bhp,5.0,NaN,1.95
10,Maruti Ciaz Zeta,Kochi,2018,25692,Petrol,Manual,First,21.56 kmpl,1462 CC,103.25 bhp,5.0,10.65 Lakh,9.95
11,Honda City 1.5 V AT Sunroof,Kolkata,2012,60000,Petrol,Automatic,First,16.8 kmpl,1497 CC,116.3 bhp,5.0,NaN,4.49


Considerations to Note:
* Just by looking at the first 5 columns, the second observation has a different unit in `Mileage` than the others (km/kg instead of kmpl)
* The 14th observation has a `Mileage` of 0

## Part B
> Note: I did this part before removing missing values (Part A)
* Remove units from attributes and only keep numerical values

Investigate the mismatch unit problem in `Mileage`:



In [21]:
# Check how many observations have km/kg
print(f'Number of rows with km/kg mileage: {cars['Mileage'].str.contains('km/kg').sum()}')

# Find rows with km/kg
cars[cars['Mileage'].str.contains('km/kg', na=False)]

Number of rows with km/kg mileage: 3


,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,13 km/kg,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
26,Nissan Micra Diesel XV,Hyderabad,2012,54000,Diesel,Manual,First,16.48 km/kg,1461 CC,63.1 bhp,5.0,NaN,4.25
58,Nissan X-Trail SLX AT,Hyderabad,2010,121812,Diesel,Automatic,First,9.49 km/kg,1995 CC,147.6 bhp,5.0,NaN,7.75


Convert the km/kg values to kmpl (multiply mileage by 1.4):

In [22]:
# Identify km/kg rows
kg_rows = cars['Mileage'].str.contains('km/kg', na=False)

# Convert km/kg to kmpl (also remove label and change float)
cars.loc[kg_rows, 'Mileage'] = cars.loc[kg_rows, 'Mileage'].str.extract(r'([\d.]+)').squeeze().astype(float) * 1.4

# Add kmpl label back so the column is consistent
cars.loc[kg_rows, 'Mileage'] = cars.loc[kg_rows, 'Mileage'].astype(str) + ' kmpl'
cars.head()

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67 kmpl,1582 CC,126.2 bhp,5.0,NaN,12.50
2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.2 kmpl,1199 CC,88.7 bhp,5.0,8.61 Lakh,4.50
3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77 kmpl,1248 CC,88.76 bhp,7.0,NaN,6.00
4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.2 kmpl,1968 CC,140.8 bhp,5.0,NaN,17.74
6,Nissan Micra Diesel XV,Jaipur,2013,86999,Diesel,Manual,First,23.08 kmpl,1461 CC,63.1 bhp,5.0,NaN,3.50


Remove unit from numerical columns:

In [23]:
# columns that have units
unit_columns = ['Mileage', 'Engine', 'Power', 'New_Price']

# keep only numerical values and converting to float type
for col in unit_columns:
  cars[col] = cars[col].str.extract(r'([\d.]+)').astype(float)

cars.head(5)

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67,1582.0,126.20,5.0,NaN,12.50
2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.20,1199.0,88.70,5.0,8.61,4.50
3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77,1248.0,88.76,7.0,NaN,6.00
4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.20,1968.0,140.80,5.0,NaN,17.74
6,Nissan Micra Diesel XV,Jaipur,2013,86999,Diesel,Manual,First,23.08,1461.0,63.10,5.0,NaN,3.50


## Part A
* Look for missing values
* Replace or drop them and justify choice.

In [24]:
# Check which columns have missing values
na_summary = pd.DataFrame({
    'missing_count': cars.isna().sum(),
    'missing_percentage': round(cars.isna().sum() / len(cars) * 100, 4)
})
na_summary

,missing_count,missing_percentage
Name,0,0.0000
Location,0,0.0000
Year,0,0.0000
Kilometers_Driven,0,0.0000
Fuel_Type,0,0.0000
Transmission,0,0.0000
Owner_Type,0,0.0000
Mileage,2,0.0342
Engine,36,0.6157
Power,36,0.6157


Observations:  
* `Mileage`, `Engine`, `Power`, and `Seats` have fairly low number of missing values
* `New_Price` has over 86% of values missing. I will drop this column.   

---
I check the observations with missing `Mileage` since there are only 2:

In [25]:
# Check which observations are missing in Mileage
cars.loc[cars['Mileage'].isna()]

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
4446,Mahindra E Verito D4,Chennai,2016,50000,Electric,Automatic,First,NaN,72.0,41.0,5.0,13.58,13.00
4904,Toyota Prius 2009-2016 Z4,Mumbai,2011,44000,Electric,Automatic,First,NaN,1798.0,73.0,5.0,NaN,12.75


The `Fuel_Type` of the two observations with missing `Mileage` values are both "Electric". So I need to check if the unique values of `Fuel_Type` to see if this is related to the missing values.

In [26]:
# Get unique values of Fuel_Type
print(f'Possible Fuel_Type values: {cars['Fuel_Type'].unique()}')

# Find where Fuel_Type is "Electric"
cars.loc[cars['Fuel_Type'] == 'Electric']

Possible Fuel_Type values: ['Diesel' 'Petrol' 'Electric']


,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
4446,Mahindra E Verito D4,Chennai,2016,50000,Electric,Automatic,First,NaN,72.0,41.0,5.0,13.58,13.00
4904,Toyota Prius 2009-2016 Z4,Mumbai,2011,44000,Electric,Automatic,First,NaN,1798.0,73.0,5.0,NaN,12.75


It's confirmed that only rows with `Fuel_Type` as "Electric" are the ones with missing values. I will remove these rows.




In [27]:
# Remove rows with missing Mileage
cars.dropna(subset=['Mileage'])

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,Diesel,Manual,First,19.67,1582.0,126.20,5.0,NaN,12.50
2,Honda Jazz V,Chennai,2011,46000,Petrol,Manual,First,18.20,1199.0,88.70,5.0,8.61,4.50
3,Maruti Ertiga VDI,Chennai,2012,87000,Diesel,Manual,First,20.77,1248.0,88.76,7.0,NaN,6.00
4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Diesel,Automatic,Second,15.20,1968.0,140.80,5.0,NaN,17.74
6,Nissan Micra Diesel XV,Jaipur,2013,86999,Diesel,Manual,First,23.08,1461.0,63.10,5.0,NaN,3.50
...,...,...,...,...,...,...,...,...,...,...,...,...,...
6014,Maruti Swift VDI,Delhi,2014,27365,Diesel,Manual,First,28.40,1248.0,74.00,5.0,7.88,4.75
6015,Hyundai Xcent 1.1 CRDi S,Jaipur,2015,100000,Diesel,Manual,First,24.40,1120.0,71.00,5.0,NaN,4.00
6016,Mahindra Xylo D4 BSIV,Jaipur,2012,55000,Diesel,Manual,Second,14.00,2498.0,112.00,8.0,NaN,2.90
6017,Maruti Wagon R VXI,Kolkata,2013,46000,Petrol,Manual,First,18.90,998.0,67.10,5.0,NaN,2.65


Investigate Mileage=0.0 problem:

In [28]:
# Find the rows that have Mileage value of 0
cars[cars['Mileage'] == 0]

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price
14,Land Rover Freelander 2 TD4 SE,Pune,2012,85000,Diesel,Automatic,Second,0.0,2179.0,115.00,5.0,NaN,17.50
67,Mercedes-Benz C-Class Progressive C 220d,Coimbatore,2019,15369,Diesel,Automatic,First,0.0,1950.0,194.00,5.0,49.14,35.67
194,Honda City 1.5 GXI,Ahmedabad,2007,60006,Petrol,Manual,First,0.0,NaN,NaN,NaN,NaN,2.95
229,Ford Figo Diesel,Bangalore,2015,70436,Diesel,Manual,First,0.0,1498.0,99.00,NaN,NaN,3.60
544,Mercedes-Benz New C-Class Progressive C 200,Kochi,2019,13190,Petrol,Automatic,First,0.0,1950.0,181.43,5.0,49.49,38.99
707,Mercedes-Benz M-Class ML 350 4Matic,Pune,2014,120000,Diesel,Automatic,First,0.0,2987.0,165.00,5.0,NaN,30.00
749,Land Rover Range Rover 3.0 D,Mumbai,2008,55001,Diesel,Automatic,Second,0.0,NaN,NaN,NaN,NaN,26.50
962,Mercedes-Benz C-Class Progressive C 220d,Mumbai,2018,8682,Diesel,Automatic,First,0.0,1950.0,194.00,5.0,52.26,39.50
996,Hyundai Santro Xing GL,Pune,2008,93000,Petrol,Manual,First,0.0,1086.0,62.00,5.0,NaN,1.45
1059,Hyundai Santro Xing GL,Hyderabad,2010,58163,Petrol,Manual,First,0.0,1086.0,62.00,5.0,NaN,2.45


There seems to be no obvious reason why the values are missing. So I change them NaN to impute them later.

In [29]:
# Replace 0.0 kmpl with NaN
cars['Mileage'] = cars['Mileage'].replace(0, np.nan)

# Check that no cars have Mileage of 0.0
cars.loc[cars['Mileage'] == 0]

,Name,Location,Year,Kilometers_Driven,Fuel_Type,Transmission,Owner_Type,Mileage,Engine,Power,Seats,New_Price,Price


Since `Seats` is a discrete variable, I want to look into count distribution:

In [30]:
# View count distribution of Seats
print(f'\n{cars['Seats'].value_counts().sort_index()}')


Seats
2.0       13
4.0       93
5.0     4866
6.0       29
7.0      668
8.0      133
9.0        3
10.0       4
Name: count, dtype: int64


Observation:
* The most frequent number of seats is 5

Next, I check the skewness of each column with missing values.

In [31]:
na_columns = ['Mileage', 'Engine', 'Power', 'Seats']

# Print skewness values of columns with missing values
for col in na_columns:
  print(f'{col}: Skewness = {cars[col].skew():.2f}')

Mileage: Skewness = 0.13
Engine: Skewness = 1.41
Power: Skewness = 1.92
Seats: Skewness = 1.91


Observations:  
* `Mileage` is approximatley symmetric
* The other columns are highly positively skewed


---

### Replace Missing Values

In [32]:
# Replace missing values in Mileage with mean
cars['Mileage'] = cars['Mileage'].fillna(cars['Mileage'].mean())

# Replace missing values in Engine with median (rounded)
cars['Engine'] = cars['Engine'].fillna(round(cars['Engine'].median(), 0))

# Replace missing values in Power with median
cars['Power'] = cars['Power'].fillna(cars['Power'].median())

# Replace missing values in Seats with mode
cars['Seats'] = cars['Seats'].fillna(cars['Seats'].mode()[0])

# Drop the New_Price column
cars.drop('New_Price', axis=1, inplace=True)

# Change Engine and Seats to integers (from floats)
cars['Engine'] = cars['Engine'].astype(int)
cars['Seats'] = cars['Seats'].astype(int)

cars.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5847 entries, 1 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5847 non-null   object 
 1   Location           5847 non-null   object 
 2   Year               5847 non-null   int64  
 3   Kilometers_Driven  5847 non-null   int64  
 4   Fuel_Type          5847 non-null   object 
 5   Transmission       5847 non-null   object 
 6   Owner_Type         5847 non-null   object 
 7   Mileage            5847 non-null   float64
 8   Engine             5847 non-null   int64  
 9   Power              5847 non-null   float64
 10  Seats              5847 non-null   int64  
 11  Price              5847 non-null   float64
dtypes: float64(3), int64(4), object(5)
memory usage: 593.8+ KB


### Justifications for Replacement/Drop
* `Mileage`: Two decisions were made for this column. The two original rows that had a missing value were found to be the only cars that had "Electric" as `Fuel_Type`. Becuase electric vehicles technically don't have mileage, these two rows were **dropped**. Rows with Mileage values = 0 were identified, converted into NaNs, then those NaN values were replaced with the **mean**, as the skewness value (0.13) shows that the distribution is symmetric. Median could also have been chosen, but since the distribution is symmetric, using the mean will preserve the original mean of the column after imputation. Also, only a small percentage missing values were present so using mean over median may not have any significant effects.
* `Engine`: missing values were replaced with **median**. The skewness value (1.41) is greater than 1, indicating a highly right-skewed distribution. In this case, median is better than mean because mean is heavily sensitive to outliers. Median better represents a typical value in the skewed data, which is the goal of imputation. I also decided to round the median value before replacement, because by visual inspection of the data and basic domain knowledge, engine displacement values are usually whole numbers.
* `Power`: missing values were replaced with **median** for the same reason as `Engine`. The only difference is that the median value is not rounded since power values are typically decimals.
* `Seats`: missing values were replaced with **mode**. While it technically is numerical data, this variable is discrete and categorical-like. There is only a small fixed set of possible values. The best assumption we can make is that cars with the missing values are probably similar to the majority of cars in the dataset. Also with basic knowledge, I know that a majority of cars have 5 seats anyway, and this is supported by the value counts. Mean or median could give a decimal value, which would not provide a real valid value.
* `New_Price`: This column was **dropped** from the dataset. ~86% of the values were missing in this column, so if I decided to impute missing values, I'd essentially be making the data up. In a real situation, the missing values of this column could be searched up online.

## Part C
* Change categorical variables into numerical one hot encoded values

In [33]:
# Encode Fuel_Type and Transmission, and convert to integers (from boolean)
cars_encoded = pd.get_dummies(cars, columns=['Fuel_Type', 'Transmission'], dtype=int)

cars_encoded.head()

,Name,Location,Year,Kilometers_Driven,Owner_Type,Mileage,Engine,Power,Seats,Price,Fuel_Type_Diesel,Fuel_Type_Electric,Fuel_Type_Petrol,Transmission_Automatic,Transmission_Manual
1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,First,19.67,1582,126.20,5,12.50,1,0,0,0,1
2,Honda Jazz V,Chennai,2011,46000,First,18.20,1199,88.70,5,4.50,0,0,1,0,1
3,Maruti Ertiga VDI,Chennai,2012,87000,First,20.77,1248,88.76,7,6.00,1,0,0,0,1
4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Second,15.20,1968,140.80,5,17.74,1,0,0,1,0
6,Nissan Micra Diesel XV,Jaipur,2013,86999,First,23.08,1461,63.10,5,3.50,1,0,0,0,1


## Part D
* Add one feature and add to the dataset
* I added a column `Car_Age` that represents the current age of the car

In [34]:
from datetime import datetime

cars_final = cars_encoded.copy()
cars_final['Car_Age'] = datetime.now().year - cars_final['Year']

cars_final.head()

,Name,Location,Year,Kilometers_Driven,Owner_Type,Mileage,Engine,Power,Seats,Price,Fuel_Type_Diesel,Fuel_Type_Electric,Fuel_Type_Petrol,Transmission_Automatic,Transmission_Manual,Car_Age
1,Hyundai Creta 1.6 CRDi SX Option,Pune,2015,41000,First,19.67,1582,126.20,5,12.50,1,0,0,0,1,11
2,Honda Jazz V,Chennai,2011,46000,First,18.20,1199,88.70,5,4.50,0,0,1,0,1,15
3,Maruti Ertiga VDI,Chennai,2012,87000,First,20.77,1248,88.76,7,6.00,1,0,0,0,1,14
4,Audi A4 New 2.0 TDI Multitronic,Coimbatore,2013,40670,Second,15.20,1968,140.80,5,17.74,1,0,0,1,0,13
6,Nissan Micra Diesel XV,Jaipur,2013,86999,First,23.08,1461,63.10,5,3.50,1,0,0,0,1,13


## Part E


### Select and Filter Operations
I used these operations to answer the question "What are the names and price of all first owner cars made after 2011?"

In [35]:
# Select and Filter operations
cars_final.loc[(cars_final['Owner_Type'] == 'First') & (cars_final['Year'] > 2011), ['Name', 'Price', 'Owner_Type', 'Year']]

,Name,Price,Owner_Type,Year
1,Hyundai Creta 1.6 CRDi SX Option,12.50,First,2015
3,Maruti Ertiga VDI,6.00,First,2012
6,Nissan Micra Diesel XV,3.50,First,2013
7,Toyota Innova Crysta 2.8 GX AT 8S,17.50,First,2016
8,Volkswagen Vento Diesel Comfortline,5.20,First,2013
...,...,...,...,...
6010,Honda Brio 1.2 VX MT,3.20,First,2013
6013,Honda Amaze VX i-DTEC,4.83,First,2015
6014,Maruti Swift VDI,4.75,First,2014
6015,Hyundai Xcent 1.1 CRDi S,4.00,First,2015


### Rename/Mutate Operations
* I renamed `Engine` to `Engine_Displacement` and `Power` to `Engine_Horsepower`
* I added a new column (mutate) `Engine_Efficiency`

In [36]:
# Rename operation
cars_final = cars_final.rename(columns={'Engine': 'Engine_Displacement', 'Power': 'Engine_Horsepower'})

# Mutate operation
cars_final['Engine_Efficiency'] = round(cars_final['Engine_Horsepower'] / cars_final['Engine_Displacement'], 4)

cars_final[['Engine_Displacement', 'Engine_Horsepower', 'Engine_Efficiency']].head()

,Engine_Displacement,Engine_Horsepower,Engine_Efficiency
1,1582,126.20,0.0798
2,1199,88.70,0.0740
3,1248,88.76,0.0711
4,1968,140.80,0.0715
6,1461,63.10,0.0432


### Arrange and Summarize with Groupby Operations
* I used a combination of these operations to answer the question " What is the average price and average kilometers driven by location, sorted by average price descending?"

In [37]:
cars_final.groupby('Location').agg(
  Avg_Price=('Price', 'mean'),
  Avg_KM=('Kilometers_Driven', 'mean')
).round(2).sort_values(by='Avg_Price', ascending=False)

,Avg_Price,Avg_KM
Location,,
Coimbatore,15.16,46949.78
Bangalore,13.48,58227.86
Kochi,11.31,44718.00
Hyderabad,10.00,70651.65
Delhi,9.88,57185.81
Mumbai,9.59,44666.16
Ahmedabad,8.57,54974.70
Chennai,7.96,90494.86
Pune,6.95,69785.16


In [38]:
# save final processed data set into a csv
cars_final.to_csv('cars_processed.csv', index=False)